platform/admin/admin-configuration-configinstaller/src/main/java/org/codice/ddf/admin/configuration/ConfigurationInstaller.java

```diff
@@ -186,61 +221,85 @@ public void handleStore(ConfigurationContext context) throws IOException {
-      LOGGER.debug("Deleting file because config was deleted for pid {}", pid);
+      LOGGER.debug("Deleting file because config was deleted for pid [{}]", pid);
       try {
         Files.delete(fileFromCache.toPath());
-        SECURITY_LOGGER.info("Removing a deleted config: {}", fileFromCache.toPath());
+        SecurityLogger.audit("Removing a deleted config [{}]", fileFromCache.toPath());
       } catch (IOException e) {
-        LOGGER.debug("Problem deleting config file: ", e);
+        LOGGER.debug(
+            format("Problem deleting config file [%s]: ", fileFromCache.getAbsolutePath()), e);
         // Synchronous with config admin, so we can report the failure to the UI this way
         throw e;
       }
     }
   }
 
-  /**
-   * In this block, based upon the circumstances of where it gets called, there's a guarantee that
-   * {@link ConfigurationContext#getConfigFile()} will not return null.
-   */
-  private void writeIfNecessary(ConfigurationContext context, CachedConfigData data)
+  private void writeIfNecessary(
+      String pid, File dest, Dictionary<String, Object> configState, CachedConfigData cachedData)
       throws IOException {
-    File dest = context.getConfigFile();
     if (dest.exists()) {
-      // Improvement: These would become checksums
-      Dictionary<String, Object> fileState = data.getProps();
-      Dictionary<String, Object> configState = context.getSanitizedProperties();
-      processPasswords(context.getServicePid(), configState, encryptionService::encryptValue);
+      // DDF-3413: Future performance/memory improvement - this would become a checksum
+      Dictionary<String, Object> fileState = cachedData.getProps();
+      processPasswords(pid, configState, this::encryptValue);
       if (!equalDictionaries(configState, fileState)) {
-        try (OutputStream outputStream = new FileOutputStream(dest)) {
+        try (OutputStream outputStream = new BufferedOutputStream(new FileOutputStream(dest))) {
           getAppropriateStrategy(dest.toPath()).write(outputStream, configState);
+        } catch (IOException e) {
+          LOGGER.debug("Writing to config file failed: ", e);
+          SecurityLogger.audit("Failure to update config file [{}]", dest.getAbsolutePath());
+          throw e;
         }
-        SECURITY_LOGGER.info("Updating config file: {}", dest.toPath());
-        data.setProps(configState);
+        SecurityLogger.audit("Updating config file [{}]", dest.toPath());
+        cachedData.setProps(configState);
       }
     }
   }
 
+  /**
+   * {@link EncryptionService} is whitelisted, so we cannot add this method onto the interface and
+   * impl until a major release.
+   */
+  private String encryptValue(String plaintextValue) {
+    Matcher m = ENC_PATTERN.matcher(plaintextValue);
+    if (m.find()) {
+      LOGGER.debug("Password already encrypted");
+      return plaintextValue;
+    }
+    return "ENC(".concat(encryptionService.encrypt(plaintextValue)).concat(")");
```

Comment:
✏️ You might also consider
`return String.format("ENC(%s)", encryptionService.encrypt(plaintextValue));`
It avoids creating an extra intermediate String.